presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushdownFilterIntoWindow.java

```diff
@@ -0,0 +1,184 @@
+import java.util.OptionalInt;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.SystemSessionProperties.isOptimizeTopNRowNumber;
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.spi.predicate.Marker.Bound.BELOW;
+import static io.prestosql.spi.predicate.Range.range;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.sql.planner.DomainTranslator.fromPredicate;
+import static io.prestosql.sql.planner.plan.Patterns.filter;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.window;
+import static java.lang.Math.toIntExact;
+
+public class PushdownFilterIntoWindow
+        implements Rule<FilterNode>
+{
+    private final Capture<WindowNode> childCapture;
+    private final Pattern<FilterNode> pattern;
+
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+    private final FunctionId rowNumberFunctionId;
+
+    public PushdownFilterIntoWindow(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.domainTranslator = new DomainTranslator(metadata);
+        this.rowNumberFunctionId = metadata.resolveFunction(QualifiedName.of("row_number"), ImmutableList.of()).getFunctionId();
+        this.childCapture = newCapture();
+        this.pattern = filter()
+                .with(source().matching(window()
+                        .matching(window -> window.getWindowFunctions().size() == 1)
+                        .matching(window -> getOnlyElement(window.getWindowFunctions().values()).getResolvedFunction().getFunctionId().equals(rowNumberFunctionId))
+                        .capturedAs(childCapture)));
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return pattern;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isOptimizeTopNRowNumber(session);
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        Session session = context.getSession();
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+        PlanNodeIdAllocator idAllocator = context.getIdAllocator();
+
+        WindowNode windowNode = captures.get(childCapture);
+
+        TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), types).getTupleDomain();
+
+        Symbol rowNumberSymbol = getOnlyElement(windowNode.getWindowFunctions().entrySet()).getKey();
+        OptionalInt upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
+
+        if (upperBound.isPresent()) {
+            if (upperBound.getAsInt() <= 0) {
+                return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
+            }
+            if (windowNode.getOrderingScheme().isPresent()) {
```

Comment:
This condition should be part of the pattern.

Reply:
If we put the condition in the pattern, the above code (L.108) will not be executed.

Reply:
Hmm, is that correct? Returning `ValuesNode` from `PushdownFilterIntoRowNumber` is still necessary if the upper bound of the filter is negative. Even if we move the code L.108 to `PushdownFilterIntoRowNumber`, `PushdownFilterIntoWindow` even needs to rewrite the `FilterNode` with a window if the upper bound is negative.

Reply:
I'm not asking you to move the code to another rule. Sorry for not making it clear.

This is a correct observation that if `windowNode.getOrderingScheme().isPresent()` becomes part of the pattern, then the case of `upperBound.getAsInt() <= 0` for a WindowNode without ordering will not be handled by this rule.
This is fine, because this case is already handled by other rules (as I mentioned previously).

By adding `windowNode.getOrderingScheme().isPresent()` to the pattern, we make a clear distinction between the rules and we don't lose any functionality.

Reply:
Umm, `PushdownFilterIntoRowNumber` matches the pattern of a filter having **row number** as a source. On the other hand, `PushdownFilterIntoWindow` matches the pattern of a filter having a **window** as a source. If we introduce `windowNode.getOrderingScheme().isPresent()` in the pattern of `PushdownFilterIntoWindow`, no one can handle the case, I think.

Reply:
For a plan
```
- Filter
     - Window (no ORDER BY)
```

First,  the `ReplaceWindowWithRowNumber ` rule will convert the WindowNode into RowNumberNode. Then, the `PushdownFilterIntoRowNumber` rule will match.